### ISSUE TITLE
discovery: monetizing github issues (incentive)

### ISSUE THREAD

**Issue by kylemann** [thnxdev/issues#8]
_TLDR? exploring idea to incentivize organizational investments in OSS work that directly impacts their projects while giving maintainers financial support in the form of donate-for-work or grant._

## Exploration

As we [explore incentives](https://github.com/thnxdev/issues/issues/7) between organizations, maintainers, and contributors in open-source, one concept is monetizing GitHub issues to create a dynamic marketplace driven by shared incentives.

For example, an organization with a critical dependency could fund a grant or bounty for a prioritized issue. Likewise, maintainers could request funding for issues that improve a project, benefiting all parties. In this setup, thanks.dev facilitates donations linked to specific dependencies, giving organizations the added value of directly supporting improvements they rely on.

![image](https://github.com/user-attachments/assets/9e59c890-bb19-4beb-aee8-310d450e323c)

---

## Concept Review

Explore thanks.dev monetizing specific GitHub issues, enabling organizations to post bounties or grants for high-priority issues or desired features. Maintainers, too, could identify issues they wish to work on that would benefit the project. To summarize:

- Organizations could fund bounties for prioritized issues, while maintainers propose tasks needing support, creating mutual incentives
- Maintainer: thanks.dev could connect high-quality contributors to funded issues, valuable contributions for projects and the community
- Contributor: has ability to earn while contributing to OSS project

📽️ [1:30 minute video review](https://youtu.be/u2aHN1vbIzQ)

**Comment by nehzata** [comment]
@kylemann This is great! We received a suggestion along similar lines from Mot (dotenv maintainer) some time ago. 
![image](https://github.com/user-attachments/assets/2c284784-af8e-411c-b955-e44818cc86c2)

## Previous attempts

These are the prior attempts at this I'm aware of that we can also learn from.
- The original idea behind polar.sh. They seem to have pivoted to payment processing but issue funding & rewards is still part of their landing page.
- algora.io which seems to be primarily focused on OSS startups using bounties to get issues fixed.

## Maintainer motivations

According to the 2022 FOSS contributor survey [Figure 9a on page 20](https://8112310.fs1.hubspotusercontent-na1.net/hubfs/8112310/2020FOSSContributorSurveyReport_121020.pdf), maintainers ranked "making an impact", "learning" and "doing creative work" higher than monetary factors as their motivations for contributing to open source.

## Our strength / differentiation point

I really like the approach you're suggesting @kylemann because thanks.dev is in a great position to align incentives here. Further to your summary:
- On the organization side, we also get to find out which projects they depends on -> which are high risk and we get to demonstrate an outcome from their sponsorship!
- On the maintainer and contributor side, your suggested approach aligns much better than the alternatives in terms of intrinsic motivations! Looking at algora, the majority of bounties are getting posted by OSS startups, whereas we're in a position to do the same for Mot et al. I'd expect the issues we'll be able to fund would be more fun and motivating.

## Next steps?

@kylemann & @vladh do you think we should continue brainstorming or should we flesh this option out more? In terms of fleshing it out, whether now or later, we can invite a few maintainers and existing TD sponsors to the conversation and validate our assumptions?

**Comment by vladh** [comment]
Thanks @kylemann, I really like the idea and I think the demo is very neat.

I don't know if/how other platforms such as [Polar](https://polar.sh/) solve this, but I wonder how we might prevent toe-stepping. Say a bounty for an issue goes up and a bunch of people start working on that issue. It might be that the only person who gets paid is the person who happens to submit a good-enough PR first. This might lead everyone else who was putting work in to feel disappointed.

To avoid this, maybe there can be an “I'm working on this” indicator. The purpose would not be to lock an issue to a single person at a time — it's okay for multiple people to work on the same thing. Rather, the aim would be transparency, so that everyone knows what's being worked on, which would avoid disappointment.

Just an idea.

Also, I do like the idea of keeping this lightweight. Most stuff can happen on GitHub, and thanks.dev can serve as an index of bountied GH issues, with minimal functionality for claiming a bounty and so on. This means we're not overcommitting to a more complicated flow than is needed.

**Comment by nehzata** [comment]
@kylemann & @vladh 
What do you think of the following marketing/distribution approach to this concept?

1. Project maintainer creates an issue with the following title format: "[RFG]: [description of task that we'd like funding for] [$BUDGET]"
2. TD's current $1m simulation knows which BigTech Co's depend on this project
3. We bring that info to the front and help the community start pinging the OSPOs of those companies to bring attention to the RFG issue and invite them to fund it via thanks.dev

@kylemann for context, TD currently scans the dependency trees of the public repositories of BigTech co's. That's how we show the $1m breakdown on the homepage.

**Comment by vladh** [comment]
@nehzata Love this idea tbh!!!

**Comment by kylemann** [comment]
@nehzata love it, too – as it leverages / ties into the existing TD workflow. 

here is some [WIP: designs](https://www.figma.com/design/fFJhf9Vinh4Qq0ENXCARey/thanks.dev?node-id=25-61&t=6OrMgyIUFvJ82deJ-1) of the potential flow. the part I could use help with is:

> We bring that info to the front and help the community start pinging the OSPOs of those companies to bring attention to the RFG issue and invite them to fund it via thanks.dev
@kylemann for context, TD currently scans the dependency trees of the public repositories of BigTech co's. That's how we show the $1m breakdown on the homepage.

perhaps we can do a quick sync to clarify / pair-design?

**Comment by vladh** [comment]
Thanks so much @kylemann, the Figma proposals are really helpful. I've added some comments directly on the proposals, happy to discuss this more. :)

**Comment by nehzata** [comment]
@kylemann sync session sounds great. let us know when is a good time for you please 🙏

**Comment by kylemann** [comment]
@vladh @nehzata sounds good, sent one for tuesday. in the interim, will go through vlad's comments [figma](https://www.figma.com/design/fFJhf9Vinh4Qq0ENXCARey/thanks.dev?node-id=25-61&t=6OrMgyIUFvJ82deJ-1)

**Comment by vladh** [comment]
Created #11 to implement a proof of concept for just the bot part.

When that's done, we could create issues to implement a proof of concept (1) in the database, (2) by creating a listing of monetised issues, etc. But the details still depend on the discussion in this issue.

**Comment by vladh** [comment]
Super basic proof of concept for the “bot” functionality:

https://github.com/user-attachments/assets/892be11c-d215-49be-a754-2218a71f2ed0

More on #11.

**Comment by kylemann** [comment]
https://resources.github.com/open-source-security-fund/

**Comment by nehzata** [comment]
Right now thanks.dev is purely an algorithmic mechanism for distributing funds among open source projects. One aspect of the above discussion is that it's suggesting our current purely algorithmic approach is limited in its ability to handle the human & community element of open source.

Based on feedback received so far, the RFGs approach requires more of the donors attention in deciding which grants to fund and how much. Donors can not be expected to have enough context on each grant to be able to optimally make decisions.

A simplified approach could be that company A allocates a budget of X onto TD with a % of that budget going into the grant pool. On the maintainer side, the same process as before is followed in order to indicate which projects have which issues that require funding.

The problem then becomes, how can we (TD) provide tooling and a mechanism for the optimal and transparent distribution of the grant pool? At a macro level, the monetization of GH issues we've discussed above is one such approach to solving this problem. Another possible approach on a very different part of the spectrum could be [this](https://opensource.com/open-organization/16/7/compensating-employees-letting-them-take-what-they-want).

Furthermore, I think this is what foundations should be doing. They instead fund events, conferences and studies for lack of a better alternative that can scale across the diversity of OSS?

The ideal solution would have the following characteristics:
- community focused
- transparent
- democratic?
- meritocratic?
- free market oriented?
- easy to understand and explain

@vladh & @kylemann wdyt?

**Comment by vladh** [comment]
I think there are two separate questions here.

The first question is — how should money be allocated to RFGs?

The second question is — how should money be allocated generally?

Regarding the first question, I think one solution is for the allocation to RFGs specifically to not go through us at all. thanks.dev pays maintainers, then maintainers decide which issues to allocate to. But then, why have RFGs to begin with? For marketing — RFGs will illustrate to companies that funding is needed, and motivate them to pay maintainers. Then maintainers can say “with support from Foocorp and Barcorp, we've managed to fill RFGs A, B, and C, which allowed us to develop important features; thanks!”.

Regarding the second question, which factors do we want to take into account when allocating money to dependencies? First:

> dependency depth

There is some logic to thanks.dev's current strategy for doing a first pass at allocation of money based on dependency depth. Direct dependencies will have been directly chosen by the company, and so are likely to offer them more value than indirect dependencies.

Now, @nehzata's points:

> meritocratic

Probably the most important metric for fund allocation, in my opinion, is value offered by a dependency to a company. If anything underlies “merit”, this is probably it. If I build a $1B company, and I use a completely minor dependency that could have been replaced by many other things in the process, it doesn't seem like I have great reason to give a big chunk of my money to that dependency. But if I use a dependency without which my business just _would not work_, it seems like I have reason to pay the maintainer of that dependency a significant amount of money. After all, their work made it possible for me to gain $1B in value, so they should have some share in the income, seeing as they have effectively done a significant part of the work. And, if a dependency that offers a lot of value stops being maintained, that is a great threat to my company, that I am incentivised to prevent by paying.

Unfortunately, only companies themselves know which dependencies are more valuable than others. On thanks.dev, you can weight funds going to individual dependencies. But the whole point is that companies do not want to put in the attention to do this.

So is there a different way to get at this information about value provided? I don't know, but it's worth thinking if we can further develop the algorithmic solution. We currently know a project's list of dependencies. But what if we did more in-depth code analysis? Let's say my project has 3 direct dependencies, `foolib`, `barlib` and `bazlib`. But say I use `barlib` extensively, in every file/module/whatever, and the others are used only a few times. Can thanks.dev algorithmically identify this, through some kind of static analysis or whatever, then weight `barlib` more highly? Even an imperfect solution seems to be an incremental improvement over only taking dependency depth into account. By doing something like this, we could take advantage of our privileged access to the company's source code, and extract as much helpful information out of it as we can, which the community cannot do, which brings me to my next point.

> transparent

Transparency seems straightforward — we publish our numbers, and publish our reasoning. But crucially, we cannot be transparent about companies' codebases. This places some serious constraints on community participation. Let's look at those.

> community focused
> democratic

In principle, I think community governance is great. But in this particular case, I see some issues. Here are some questions:

* When a company signs up to thanks.dev, they do so while being aware ahead of time approximately how their funds will be distributed, ie aware of the thanks.dev algorithm. Say we let the community decide how funds are allocated. Will companies be unhappy if the community decides to allocate funds very differently than the company would have?
* A company's dependencies will be public, but the company's source code will not. So community members have no way of knowing how important each dependency is to that company's codebase. Regarding merit above — what if the community decides to allocate funds to dependencies in a way that greatly mismatches the value offered to a company by their dependencies? So minor things get a lot of money, major things get nothing.
* Will allowing the community to decide the allocation turn things into a popularity contest, where technologically crucial, but unpopular, dependencies get nothing, though they should get a lot?
* If the community (partially) decides allocation, what level of expertise is required to participate? Can non-programmers vote? Can only programmers vote? Can only a board of highly qualified programmers vote?

> easy to understand and explain

I agree — this is a step we should take _after_ we ourselves have figured out the best allocation.

> free market oriented?

I don't think markets are compatible with Open Source software. The defining feature of a market is withholding goods based on mutual disinterest [^1]. For us to engage in a market transaction in a situation where you want my software, I must _withhold_ my software until you are willing to pay a sufficient price. Of course, I could just give you the software no matter how much/little you pay, or for free because you really need it. But this is no longer a market transaction (see Hussain pages below for a more thorough explanation). So I think there is a fundamental mismatch between Open Source software and markets that we should keep in mind. And I think that's okay — not everything needs to be market-based, and many people think markets inhibit us from caring about each other [^2].

[^1]: Waheed Hussain, “Living with the Invisible Hand: Markets, Corporations, and Human Freedom” (2023), sec A.2.5. https://vlad.website/t/hussain-2023-p-211-to-213.pdf
[^2]: Barry Maguire, “Efficient Markets and Alienation”; John O'Neill, “The Market”.

**Comment by nehzata** [comment]
very valid points to consider @vladh ... more thinking required

**Comment by vladh** [comment]
Regarding the point about using RFGs as marketing — could we integrate with Open Collective to show project budgets, in order to motivate companies to donate?

[BackYourStack][bys] does this (see “Yearly budget”):

![IMG_7349](https://github.com/user-attachments/assets/5db869cd-4482-44c4-b6e7-4d0c89695ae1)

[bys]: https://backyourstack.com/
